File: src/main/java/com/koddy/server/global/log/LoggingStatusManager.java
package com.koddy.server.global.log;

import org.slf4j.MDC;
import org.springframework.stereotype.Component;

import static com.koddy.server.global.filter.MdcKey.REQUEST_ID;

@Component
public class LoggingStatusManager {
    private final ThreadLocal<LoggingStatus> statusContainer = new ThreadLocal<>();

    public void syncStatus() {
        final LoggingStatus status = statusContainer.get();
        if (status == null) {
            final LoggingStatus firstLoggingStatus = new LoggingStatus(MDC.get(REQUEST_ID.name()));
            statusContainer.set(firstLoggingStatus);
        }
    }

    public LoggingStatus getExistLoggingStatus() {
        final LoggingStatus status = statusContainer.get();
        if (status == null) {
            throw new IllegalStateException("ThreadLocal LoggingStatus not exists...");
        }
        return status;
    }

    public void increaseDepth() {
        statusContainer.get().increaseDepth();
    }

    public void decreaseDepth() {
        statusContainer.get().decreaseDepth();
    }

    public void clearResource() {
        statusContainer.remove();
    }
}

File: src/main/java/com/koddy/server/global/log/LoggingStatusManager.kt
package com.koddy.server.global.log

import org.springframework.stereotype.Component

@Component
class LoggingStatusManager {
    private val statusContainer = ThreadLocal<LoggingStatus>()

    fun getExistLoggingStatus(): LoggingStatus =
        statusContainer.get()
            ?: throw IllegalStateException("ThreadLocal LoggingStatus not exists...")

    fun syncStatus() {
        val status: LoggingStatus? = statusContainer.get()
        if (status == null) {
            statusContainer.set(LoggingStatus())
        }
    }

    fun clearResource() = statusContainer.remove()
}
